controller/src/main/java/io/pravega/controller/task/Stream/StreamMetadataTasks.java

```diff
@@ -210,7 +256,217 @@ public void initializeStreamWriters(final EventStreamClientFactory clientFactory
+                        log.debug("ReaderGroup Config is {}", rgConfigRecord.getObject().isUpdating());
+                        return CompletableFuture.completedFuture(!rgConfigRecord.getObject().isUpdating());
+                   });
+    }
+
+    /**
+     * Delete Reader Group.
+     *
+     * @param scope           Reader Group scope.
+     * @param rgName          Reader Group name.
+     * @param readerGroupId   Reader Group unique identifier.
+     * @param contextOpt      Reader Group context.
+     * @return deletion status.
+     */
+    public CompletableFuture<DeleteReaderGroupStatus.Status> deleteReaderGroup(final String scope, final String rgName,
+                                                                               final String readerGroupId,
+                                                                               RGOperationContext contextOpt) {
+        final RGOperationContext context = contextOpt == null ? streamMetadataStore.createRGContext(scope, rgName) : contextOpt;
+        final long requestId = requestTracker.getRequestIdFor("deleteReaderGroup", scope, rgName);
+        return RetryHelper.withRetriesAsync(() -> streamMetadataStore.getReaderGroupId(scope, rgName, context, executor)
+                .thenCompose(rgId -> {
+                    if (!rgId.equals(UUID.fromString(readerGroupId))) {
+                        return CompletableFuture.completedFuture(DeleteReaderGroupStatus.Status.RG_NOT_FOUND);
+                    }
+                   return streamMetadataStore.getVersionedReaderGroupState(scope, rgName, true, context, executor)
+                            .thenCompose(versionedState -> eventHelper.addIndexAndSubmitTask(new DeleteReaderGroupEvent(scope, rgName, requestId, rgId),
+                                      () -> startReaderGroupDelete(scope, rgName, versionedState, context))
+                                      .thenCompose(x -> eventHelper.checkDone(() -> isRGDeleted(scope, rgName))
+                                      .thenApply(done -> DeleteReaderGroupStatus.Status.SUCCESS)));
+                }), e -> Exceptions.unwrap(e) instanceof RetryableException, READER_GROUP_OPERATION_MAX_RETRIES, executor);
+    }
+
+    private CompletableFuture<Boolean> isRGDeleted(String scope, String rgName) {
+        return streamMetadataStore.checkReaderGroupExists(scope, rgName).thenApply(exists -> !exists);
+    }
+
+    private CompletableFuture<Void> startReaderGroupDelete(final String scope, final String rgName, VersionedMetadata<ReaderGroupState> currentState,
+                                                           RGOperationContext context) {
+        final String rgStreamName = NameUtils.getStreamForReaderGroup(rgName);
+        return Futures.toVoid(streamMetadataStore.updateReaderGroupVersionedState(scope, rgName,
+           ReaderGroupState.DELETING, currentState, context, executor))
+           .thenCompose(v -> sealStream(scope, rgStreamName, null))
```

Comment:
the guarantee of eventhelper.addIndexAndSubmitTask is for one update. 
a failing over of this will simply post the event. it doesnt have the inputs for the future. 
so if you actually do have intent to do multiple updates, they should be done as part of event processing and not here. 
here we should just to do the fist step, start the operation and then post the event for it which will perform the whole execution if the processing has been started.

here you are doing  `updateReaderGroupVersionedState` followed by `sealStream` which in itself does a bunch of things. 
i suggest you move seal stream inside the event processing. but that is a different beast all together. but we have a new issue to take one async processing waiting on another out of the workflow.. we will handle that later. for now, simply move this there.

Reply:
sure. I'll move seal stream inside eventprocessing